Feature/implement history filter

jaywon opened the pull request:
Main functionality to resolve #233 :

* Sticky header for field titles and filter
* Support for both in string filter matches and regular expressions, including visual invalid RegEx state styling and toggle
* Styling for smaller font-size and table element spacing
* Filtered/total message count awareness
* Bug fix for document.getElementById error when no messages are visible with filter when targeting last message element as new messages load

psiinon commented:
Thanks for this :) Will review as soon as I get a chance...

psiinon commented on src/main/zapHomeFiles/hud/drawer.html:
```diff
@@ -20,24 +21,50 @@
+                            </span>
+                        </th>
+                        <th class="title hidden">
+                            <span>{{ hiddenMessageCount }} of {{ messageCount }} items hidden by filters</span>
```
Should be i18n'd

psiinon commented on src/main/zapHomeFiles/hud/drawer.html:
```diff
@@ -20,24 +21,50 @@
+                                <input v-model="regexEnabled" type="checkbox" />
+                            </span>
+                            <span>
+                                Enable RegEx
```
Should be i18n'd

jaywon commented:
Let me know. I had pushed and opened PR to the private repo before the multi-tab support got merged to `develop` and repo switched. After that got merged and merging into my fork i wasn't seeing history messages being dispatched at all, and haven't confirmed yet if it was an issue w/ my environment or something about the merge. 

Will try to get some time into it this weekend to troubleshoot further. Also want to start looking at tests around the API for verifying these kinds of things.

jaywon commented:
@psiinon Yup! Will make those changes and I did want to reduce the font size and add a draggable bar to let the user manage visibility based on personal preference. 👍

psiinon commented:
We should probably have a 'global' control to manage font size, as this is something that should apply to all tools.

dscrobonia commented on src/main/zapHomeFiles/hud/drawer.js:
```diff
@@ -12,19 +12,57 @@ Vue.component('history', {
+                    re = new RegExp(this.filter);
+                }
+                catch (ex) {
+                    re = new RegExp(''); //hack for now to prevent errors being thrown for invalid RegExp
```
what would be the "non hack" solution? would it be to also alert the user of invalid regex?

jaywon replied:
I was thinking the "non hack" solution would be to add some visual indicator to the user that the current RegEx was invalid, like a red outline around the input box? Would be easy to wire up to a state in this `catch` as they type out and can easily understand why no messages are showing. Kind of thinking how wireshark changes from green background to red as your typing in filters but more subtle. What you think?

I could do it on this PR, was really trying to get this in before launch but since I missed that I don't mind taking a little more time for polish.

dscrobonia replied:
I think that's a great idea! if you can add that in without too much more effort I think that'd be a great addition. :)

jaywon replied:
Implemented in latest revisions!

psiinon commented:
I quite like the hidden messages counter ;)

jaywon commented:
Gonna make those changes today. I actually took the hidden messages concept from Firefox Console tab filter 😄

jaywon commented:
@thc202 Sure! Can you clarify where you'd like to see that?  The description on this PR or is there somewhere else you'd like this documented?

thc202 commented:
In [CHANGELOG.md](https://github.com/zaproxy/zap-hud/blob/develop/CHANGELOG.md), worth rebase/merge the latest changes to avoid conflicts.

dscrobonia commented:
Do we have notes somewhere about how to use the changelog? this is new to me.

psiinon commented:
btw I'm raised my previous comment about font sizes as a separate issue #355

jaywon commented:
@psiinon Lemme look into that bug. I made a few more small UI/UX tweaks and I think w/ that bug squashed this should address all feedback and hopefully in a good merge state!

jaywon commented:
@psiinon @dscrobonia @thc202 Think I addressed all feedback from the team including addition to CHANGELOG, styling issues, latent count display, and i18n support. Still tweaks to be made but I'm hoping that this is in a good enough state to merge and create new PR(s) for additional tweaks.

psiinon commented on src/main/zapHomeFiles/hud/drawer.html:
```diff
@@ -20,24 +21,50 @@
+                            </span>
+                        </th>
+                        <th class="title hidden">
+                            <span>{{ hiddenMessageCount }} {{ commonOf }} {{ messageCount }} {{ historyItemsFilteredSuffix }}</span>
```
Just noticed this - we're assuming the text structure is "x of y etc" which is obviously right in English but might not be right in other languages, eg Japanese / Chinese etc? 
I think the solution is to have a parameterized i18n string like: 
history_of = {0} of {1} items hidden by filters
and to pass the values as params

jaywon replied:
Good catch! I'll take care of that today 👍

jaywon replied:
@psiinon Updated w/ latest request!

thc202 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+ - Add plain text/regex filtering capability to History section [#286](https://github.com/zaproxy/zap-hud/issues/233)
```
This should be added under the `Added` section below.

thc202 commented on src/main/zapHomeFiles/hud/drawer.js:
```diff
@@ -12,19 +12,73 @@ Vue.component('history', {
+        },
+        historyItemsFiltered() {
+            let historyItemsFilteredString = I18n.t("history_items_filtered")
+                    .replace('{0}', this.hiddenMessageCount)
```
With latest changes (merge upstream) these can be passed as parameters to `t`, like in:
https://kazupon.github.io/vue-i18n/guide/formatting.html#list-formatting

jaywon replied:
Validated that works! I had tried that method previously but the interpolation wasn't working before. Merging upstream fixed, thanks for heads up, felt wrong. 👍

thc202 commented on src/main/zapHomeFiles/hud/drawer.js:
```diff
@@ -12,19 +12,73 @@ Vue.component('history', {
+                this.hiddenMessageCount = (!this.messages) ? 0 : this.messages.length - visibleMessages.length;
+            })
+        },
+        regexEnabled(){
```
Changing just the regex state does not affect the message that shows the number of items filtered, it keeps showing the old state even though the messages are filtered.

jaywon replied:
@thc202 nice catch! I refactored for the reactivity to trigger on the filtered message list rather than the filter input itself which made for a better solution of handling the various use cases.

thc202 commented on src/other/resources/UIMessages/UIMessages.properties:
```diff
@@ -59,6 +59,7 @@ common_start = Start
 common_stop = Stop
 common_turn_on = Turn On
 common_turn_off = Turn Off
+common_of = of
```
Remove? Does not seem to be in use anymore.

thc202 commented:
Thanks!
